nh.c

```diff
@@ -461,3 +467,57 @@ void nh_failsafe_uninstall(nh_failsafe_t *fs) {
     nh_log("(NickelHook) ... failsafe: info: deleting %s", fs->tmp);
     unlink(fs->tmp);
 }
+
+// --- File/directory deletion
+
+// The following blacklist helps prevent any whoopsies when deleting files or directories during uninstall.
+
+// Let's not allow deleting any files in /bin, /sbin, /etc/init.d or u-boot stuff.
+static const char *delete_prefix_blacklist[] = {
+    "/bin",
+    "/sbin",
+    // note, '/etc/u-boot/*/u-boot.recovery' is a critical file for recovery
+    "/etc/u-boot",
+    "/etc/init.d",
+    "/usr/local/Kobo/pickel",
+};
+
+// nh_delete_path deletes the provided path if it exists. It performs some checks
+// to detect and avoid deleting critical system files
+// 
+// This is for the schmuck who forgot to terminate his array of paths
+// and accidentally deleted '/bin/sh' due to an out-of-bounds read.
+static bool nh_delete_path(const char *path, bool is_dir) {
+    if (!path) {
+        nh_log("(NickelHook) no path supplied");
+        return false;
+    }
+    char fn[PATH_MAX] = {0};
+    if (!realpath(path, fn)) {
+        nh_log("(NickelHook) unable to get canonical path for %s : %m", path);
+        return false;
+    }
+    for (size_t i = 0; i < (sizeof(delete_prefix_blacklist) / sizeof(*delete_prefix_blacklist)); i++) {
+        if (!strncmp(delete_prefix_blacklist[i], fn, strlen(delete_prefix_blacklist[i]))) {
+            nh_log("(NickelHook) not deleting %s with blacklisted prefix %s", fn, delete_prefix_blacklist[i]);
+            return false;
+        }
+    }
+    nh_log("(NickelHook) ... deleting %s %s", (is_dir ? "directory" : "file"), fn);
+    int res = is_dir 
+        ? rmdir(fn) 
+        : unlink(fn);
+    if (res == -1) {
+        nh_log("(NickelHook) failed to delete %s, with error: %m", fn);
+        return false;
```

Comment:
One more thing: I would probably return true if errno is ENOENT (check it before nh_log), as the point of the function is to delete the file, so it is essentially successful if the file doesn't exist anymore.

Reply:
Good idea. The simplest thing to do, is to extend `if (res == -1)` to `if (res == -1 && errno != ENOENT)`, but it doesn't log that case. Are you happy with that? or would you prefer the 'not exist' case be logged as well?

Reply:
Yes, that should be good.